src/algebra/module/graded_module.lean

```diff
@@ -0,0 +1,103 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import ring_theory.graded_algebra.basic
+import algebra.direct_sum.decomposition
+
+/-!
+# Graded Module
+
+Given an `R`-algebra `A` graded by `𝓐`, a graded `A`-module `M` is expressed as
+`direct_sum.decomposition 𝓜` and `set_like.has_graded_smul 𝓐 𝓜`.
+Then `⨁ i, 𝓜 i` is an `A`-module and is isomorphic to `M`.
+
+## Tags
+
+graded module
+-/
+
+
+open_locale direct_sum big_operators
+
+variables {ι R A M σ : Type*}
+variables [decidable_eq ι] [add_monoid ι] [comm_semiring R] [semiring A] [algebra R A]
+variables (𝓐 : ι → submodule R A)
+variables [add_comm_monoid M] [module A M]
+variables [set_like σ M] [add_submonoid_class σ M] (𝓜 : ι → σ)
+
+namespace graded_module
+
+instance graded_algebra.to_graded_module [graded_algebra 𝓐] :
+  set_like.has_graded_smul 𝓐 (λ i, (𝓐 i).to_add_submonoid) :=
+{ smul_mem := λ i j x y hi hj, set_like.graded_monoid.mul_mem hi hj }
+
+/--
+`⨁ i, 𝓜 i` is also an `A`-module, via `a • z = decompose (a • redecompose z)` where `decompose` and
+`recompose` are the cannonical homomorphism `M → ⨁ i, 𝓜 i` and `⨁ i, 𝓜 i → M`.
+-/
+def is_module [graded_algebra 𝓐] [@direct_sum.decomposition ι M σ _ _ _ _ 𝓜] :
```

Comment:
Is the `@` really needed here?

Reply:
Yes, Lean struggles to find `add_submonoid_class σ ?m_1` and `set_like σ ?m_1`.